src/useMutationObserver.ts

```diff
@@ -0,0 +1,68 @@
+import useCustomEffect, { IsEqual } from './useCustomEffect'
+import isEqual from 'lodash/isEqual'
+import useImmediateUpdateEffect from './useImmediateUpdateEffect'
+import useMountEffect from './useMountEffect'
+import useEventCallback from './useEventCallback'
+import { useRef } from 'react'
+
+type Deps = [Element | null | undefined, MutationObserverInit]
+
+const isDepsEqual: IsEqual<Deps> = (prev, next) =>
+  prev[0] === next[0] && isEqual(prev[1], next[1])
+
+/**
+ * Observe mutations on a DOM node or tree of DOM nodes.
+ * Depends on the `MutationObserver` api.
+ *
+ * ```ts
+ * const [element, attachRef] = useCallbackRef(null);
+ *
+ * useMutationObserver(element, { subtree: true }, (records) => {
+ *
+ * });
+ *
+ * return (
+ *   <div ref={attachRef} />
+ * )
+ * ```
+ *
+ * @param element The DOM element to observe
+ * @param config The observer configuration
+ * @param callback A callback fired when a mutation occurs
+ */
+function useMutationObserver(
+  element: Element | null | undefined,
+  config: MutationObserverInit,
+  callback: MutationCallback,
+): void {
+  const observerRef = useRef<MutationObserver | null>()
+  const fn = useEventCallback(callback)
+
+  useMountEffect(() => {
+    if (!element) return
+
+    observerRef.current = new MutationObserver(fn)
+  })
+
+  useCustomEffect(
+    () => {
+      if (!element) return
+
+      const observer = observerRef.current || new MutationObserver(fn)
+      observer.observe(element, config)
+
+      return () => {
+        observer.disconnect()
+      }
+    },
+    [element, config],
+    {
+      isEqual: isDepsEqual,
+      // Intentionally done in render, otherwise observer will miss any
+      // changes made to the DOM during this update
+      effectHook: useImmediateUpdateEffect,
```

Comment:
but this won't run on first mount, then, will it?

Reply:
i.e. how does `.observe` get called for the initial mount?

also, i don't quite understand how this is supposed to work with a normal ref.

if i pass in `elementRef.current` as the initial arg, then this won't "see" the update, will it? things won't re-render, &c.

this only works with `useCallbackRef` because that happens to force a re-render... which actually now that i look at it is sort of odd, no? because those aren't the semantics for callback refs normally...

Reply:
> because those aren't the semantics for callback refs normally...

In the sense that in class components they don't trigger updates yes, but the hook is specifically for triggering a change when the element is set or updated. You can't pas in `elementRef.current` to another hook and expect it to work generally. 

If we wanted to accept mutable element refs, i could do that with the custom hook, but we generally have only accepted actual elements as parameters for these sort of hooks since it's simpler.